app/invocation/invocation_error_card.tsx

```diff
@@ -3,75 +3,174 @@ import React from "react";
+    for (const event of props.model.aborted) {
+      if (!event.aborted) continue;
+      model.errors.push({ aborted: event });
+    }
+    if (props.model.finished?.failureDetail?.message) {
+      model.errors.push({ finished: props.model.finished });
+    }
+    return model;
   }
 
-  render() {
-    let title = "";
-    let description = "";
-    if (this.props.model.failedAction?.action?.failureDetail?.message) {
-      title = `${this.props.model.failedAction?.action?.type} action failed with exit code ${this.props.model.failedAction?.action?.exitCode}`;
-      description = this.props.model.failedAction.action.failureDetail.message;
-      const spawnCode = this.props.model.failedAction?.action?.failureDetail?.spawn?.code;
-      if (spawnCode) {
-        title += ` (${failure_details.Spawn.Code[spawnCode]})`;
+  inFlightFetch: CancelablePromise | null = null;
+  fetchModelData(model: CardModel) {
+    this.inFlightFetch?.cancel();
+
+    const promises: Promise<any>[] = [];
+    for (const error of model.errors) {
+      if (error.action?.stderr?.uri) {
+        promises.push(
+          rpc_service
+            .fetchBytestreamFile(error.action?.stderr?.uri, this.props.model.getInvocationId(), "text")
+            .then((stderr) => {
+              error.actionStderr = stderr;
+            })
+            .catch((e) => console.error("Failed to fetch failed action stderr:", e))
+        );
+      }
+      if (error.action?.stdout?.uri) {
+        promises.push(
+          rpc_service
+            .fetchBytestreamFile(error.action?.stdout?.uri, this.props.model.getInvocationId(), "text")
+            .then((stdout) => {
+              error.actionStdout = stdout;
+            })
+            .catch((e) => console.error("Failed to fetch failed action stdout:", e))
+        );
       }
-    } else if (this.props.model.aborted?.aborted?.description) {
-      title = "Build aborted";
-      description = this.props.model.aborted.aborted.description;
-    } else {
-      return null;
     }
 
-    // De-emphasize the "--sandbox_debug" message.
-    description = description.replaceAll(debugMessage, `\x1b[90m${debugMessage}\x1b[0m\n \n`);
+    this.inFlightFetch = new CancelablePromise(Promise.all(promises));
+    this.inFlightFetch.then(() => this.setState({ model }));
+  }
+
+  getTitle(model: CardModel) {
+    // If there was a failed action with a non-zero exit code, use that as the title.
+    for (const error of model.errors) {
+      if ((error.action?.exitCode ?? 0) !== 0) {
+        return `${this.props.model.failedAction?.action?.type} action failed with exit code ${this.props.model.failedAction?.action?.exitCode}`;
+      }
+    }
+    // If the finished event exists and there's a non-zero exit code, use that as the card title.
+    for (const error of model.errors) {
+      if ((error.finished?.exitCode ?? 0) !== 0) {
+        return exitCode(error.finished?.exitCode?.name ?? "");
```

Comment:
i think all of the wildcards on the return here are not needed and `name` is defined, right?